packages/opencensus-resource-util/src/resource-utils.ts

```diff
@@ -0,0 +1,204 @@
+/**
+ * Copyright 2019, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Utilities for for auto detecting monitored resource based on the environment
+ * where the application is running.
+ */
+
+import {Labels, Resource} from '@opencensus/core';
+import * as gcpMetadata from 'gcp-metadata';
+import * as http from 'http';
+import * as os from 'os';
+import * as constants from './resource-labels';
+
+export const AWS_INSTANCE_IDENTITY_DOCUMENT_URI =
+    'http://169.254.169.254/latest/dynamic/instance-identity/document';
+let resourceType: ResourceType|undefined;
+let gkeResourceLabels: Labels = {};
+let gceResourceLabels: Labels = {};
+let awsResourceLabels: Labels = {};
+
+/** Currently supported monitored resource types */
+export enum ResourceType {
+  GCP_GKE_CONTAINER = 'k8s_container',
+  GCP_GCE_INSTANCE = 'gce_instance',
+  AWS_EC2_INSTANCE = 'aws_ec2_instance',
+  NONE = 'NONE'
+}
+
+/** Determine the compute environment in which the code is running. */
+export async function getResourceType(): Promise<ResourceType> {
+  if (!resourceType) {
```

Comment:
Optional nit: what would you think about making this first line to be `if (resourceType) return resourceType;`, which would allow if statements below to be less nested.